File: python/be/contract/chain/units.py
"""Decimal strings to contract units, exactly -- never via a float.
Mirror of node/be/contract/chain/units.js."""

from __future__ import annotations

import re
from decimal import Decimal


def to_units(text, decimals: int) -> int:
    s = str(text).strip().replace("$", "").replace(",", "").replace(" ", "")
    m = re.match(r"^(-)?(\d+)(?:\.(\d+))?$", s)
    if not m:
        # pytest-bdd hands floats for {float}; render them exactly through Decimal
        d = Decimal(str(text))
        s = format(d, "f")
        m = re.match(r"^(-)?(\d+)(?:\.(\d+))?$", s)
        if not m:
            raise ValueError("not a decimal: " + str(text))
    frac = (m.group(3) or "").ljust(decimals, "0")
    if len(frac) > decimals:
        raise ValueError(f"{text} has more than {decimals} decimals")
    n = int(m.group(2) + frac)
    return -n if m.group(1) else n


def from_units(n: int, decimals: int) -> str:
    neg = n < 0
    s = str(-n if neg else n).rjust(decimals + 1, "0")
    int_part = s[: len(s) - decimals]
    frac = s[len(s) - decimals :].rstrip("0")
    return ("-" if neg else "") + int_part + ("." + frac if frac else "")


def price(t) -> int:
    return to_units(t, 8)


def size(t) -> int:
    return to_units(t, 8)


def usd(t) -> int:
    return to_units(t, 6)


def show_price(n: int) -> str:
    return from_units(n, 8)


def show_size(n: int) -> str:
    return from_units(n, 8)


def show_usd(n: int) -> str:
    return from_units(n, 6)


def notional(size_units: int, price_units: int) -> int:
    return (size_units * price_units) // 10_000_000_000

File: node/be/contract/chain/units.js
'use strict';

/**
 * Decimal strings to contract units, exactly. "2502.5" at 8 decimals is
 * 250250000000n -- never via a float, which would turn 0.1 into
 * 0.1000000000000000055.
 */
function toUnits(text, decimals) {
  const s = String(text).trim().replace(/[$,\s]/g, '');
  const m = s.match(/^(-)?(\d+)(?:\.(\d+))?$/);
  if (!m) throw new Error('not a decimal: ' + text);
  const frac = (m[3] || '').padEnd(decimals, '0');
  if (frac.length > decimals) throw new Error(`${text} has more than ${decimals} decimals`);
  const n = BigInt(m[2] + frac);
  return m[1] ? -n : n;
}

function fromUnits(n, decimals) {
  const neg = n < 0n;
  const s = (neg ? -n : n).toString().padStart(decimals + 1, '0');
  const int = s.slice(0, s.length - decimals);
  const frac = s.slice(s.length - decimals).replace(/0+$/, '');
  return (neg ? '-' : '') + int + (frac ? '.' + frac : '');
}

const price = (t) => toUnits(t, 8);
const size = (t) => toUnits(t, 8);
const usd = (t) => toUnits(t, 6);
const showPrice = (n) => fromUnits(n, 8);
const showSize = (n) => fromUnits(n, 8);
const showUsd = (n) => fromUnits(n, 6);

/** USDC notional of size at price, as the contract computes it. */
function notional(sizeUnits, priceUnits) {
  return (sizeUnits * priceUnits) / 10_000_000_000n;
}

module.exports = { toUnits, fromUnits, price, size, usd, showPrice, showSize, showUsd, notional };
